### PR TITLE
Let Sidekiq and Que set provider_job_id

### LINKED ISSUE #18821: ActiveJob not returning true job_id
ActiveJob does not return provider's job_id but uuid instead.

I use `Que` which stores an incremental id for each job (as a numeric value) in database. When executing `MyJob.perform_later(MyRecord.first)` the returning `job_id` is not the one from the database but ActiveJob's unique value instead. The `job_id` should be the one from the database in this case.

Ref: https://github.com/chanks/que/issues/84

I mad a workaround for now: 
```ruby
# config/initializers/active_job_fix.rb
class ActiveJob::Base
  def enqueue(options={})
    self.scheduled_at = options[:wait].seconds.from_now.to_f if options[:wait]
    self.scheduled_at = options[:wait_until].to_f if options[:wait_until]
    self.queue_name   = self.class.queue_name_from_part(options[:queue]) if options[:queue]
    # hide this to set job_id before logger picks it (I would change ActiveJob::Logger before_enqueue to after_enqueue)
    # run_callbacks :enqueue do
      if self.scheduled_at
        j = self.class.queue_adapter.enqueue_at self, self.scheduled_at
      else
        j = self.class.queue_adapter.enqueue self
      end
      self.job_id = j.attrs["job_id"] || j.attrs["id"] # this should be set inside each adapter
    # end
    self
  end
end
```

### PR DESCRIPTION
**Pull request by jvanbaarsen**
When a job is added to Sidekiq or Que by ActiveJob, make sure we still can get the
original job_id provider by Sidekiq or Que. We do this by adding the Sidekiq jid or Que job_id to
provider_job_id field on the job object.

Partly fixes #18821

This PR is an extension on the PR submitted by @kddeisz : https://github.com/rails/rails/pull/19910

### CODE REVIEW

**Comment by mperham** [comment]
Thank you!  This is a pretty critical integration point with Sidekiq since all its APIs require the JID to work.

**Comment by jvanbaarsen** [comment]
Added Que job_id reporting as well.

**Comment by kddeisz** [comment]
Nice work @jvanbaarsen

**Comment by rafaelfranca** [comment]
@jvanbaarsen could you rebase your branch. Thanks

**Comment by jvanbaarsen** [comment]
@rafaelfranca Yes! will do

**Comment by kddeisz** [comment]
You might also want to make your adapter tests use the same test as the
delayed job one now - and just check for not null. We don't need both tests.

On Thu, May 7, 2015 at 3:48 PM, Jeroen van Baarsen <notifications@github.com
> wrote:

> @rafaelfranca <https://github.com/rafaelfranca> Yes! will do
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20056#issuecomment-99994061>.
>



-- 
*Kevin D. Deisz*
*TrialNetworks* - part of DrugDev
Software Developer
383 Elliot Street, Suite G
Newton, MA 02464
+1 617.952.4071 x134 (office)
+1 703.615.0396 (mobile)
kdeisz@trialnetworks.com

**Comment by jvanbaarsen** [comment]
@kddeisz Ah yeah, good point! Will do as well

**Comment by jvanbaarsen** [comment]
@rafaelfranca You want me to squash the commits as well?

**Comment by rafaelfranca** [comment]
It is fine these two commits

File: activejob/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+*   Allow `Sidekiq` and `que` to report the job id back to `ActiveJob::Base` as
```
Comment (rafaelfranca) [inline activejob/CHANGELOG.md]: Instead of adding a new entry maybe we should change the entry bellow. WDYT?

Comment (jvanbaarsen) [inline activejob/CHANGELOG.md]: @rafaelfranca Yes sounds good.

Comment (kddeisz) [inline activejob/CHANGELOG.md]: Seems fine

File: activejob/CHANGELOG.md [hunk 2]
```diff
@@ -1,9 +1,9 @@
-*   Allow `DelayedJob` to report id back to `ActiveJob::Base` as
-    `provider_job_id`.
+*   Allow `DelayedJob`, `Sidekiq` and `que` to report the job id back to
+    `ActiveJob::Base` as `provider_job_id`.
 
     Fixes #18821.
 
-    *Kevin Deisz*
+    *Kevin Deisz* And *Jeroen van Baarsen*
```
Comment (jvanbaarsen) [inline activejob/CHANGELOG.md]: @kddeisz you agree with this?

File: activejob/CHANGELOG.md (hunk 1 above)
Comment (kddeisz) [inline activejob/CHANGELOG.md]: While we're here, are there any other adapters that need this? resque?

On Thu, May 7, 2015 at 3:58 PM, Jeroen van Baarsen <notifications@github.com
> wrote:

> In activejob/CHANGELOG.md
> <https://github.com/rails/rails/pull/20056#discussion_r29889786>:
>
> > @@ -1,3 +1,8 @@
> > +*   Allow `Sidekiq` and `que` to report the job id back to `ActiveJob::Base` as
>
> @rafaelfranca <https://github.com/rafaelfranca> Yes sounds good.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20056/files#r29889786>.
>



-- 
*Kevin D. Deisz*
*TrialNetworks* - part of DrugDev
Software Developer
383 Elliot Street, Suite G
Newton, MA 02464
+1 617.952.4071 x134 (office)
+1 703.615.0396 (mobile)
kdeisz@trialnetworks.com

File: activejob/CHANGELOG.md (hunk 2 above)
Comment (kddeisz) [inline activejob/CHANGELOG.md]: I'm fine with that!

On Thu, May 7, 2015 at 4:00 PM, Jeroen van Baarsen <notifications@github.com
> wrote:

> In activejob/CHANGELOG.md
> <https://github.com/rails/rails/pull/20056#discussion_r29889974>:
>
> >
> >      Fixes #18821.
> >
> > -    *Kevin Deisz*
> > +    *Kevin Deisz* And *Jeroen van Baarsen*
>
> @kddeisz <https://github.com/kddeisz> you agree with this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/20056/files#r29889974>.
>



-- 
*Kevin D. Deisz*
*TrialNetworks* - part of DrugDev
Software Developer
383 Elliot Street, Suite G
Newton, MA 02464
+1 617.952.4071 x134 (office)
+1 703.615.0396 (mobile)
kdeisz@trialnetworks.com

**Comment by jvanbaarsen** [comment]
@rafaelfranca I think its ready to be merged (Once the tests are good :dancer: )

**Comment by mperham** [comment]
Will this get into 4.2 or 5+ only?

**Comment by rafaelfranca** [comment]
5+ only.

**Comment by diddeb** [comment]
@mperham So, in current versions there's no way to find the underlying `jid` from Sidekiq using ActiveJob?

**Comment by mperham** [comment]
Not that I'm aware of.

**Comment by jvanbaarsen** [comment]
@diddeb Not out of the box. Only way you can achieve this, is by monkeypatching.

**Comment by diddeb** [comment]
Thanks @jvanbaarsen + @mperham... Freedom Patching it is.

**Pull request merged**